### PR TITLE
[Ong Cheng Seong] iP

### PR DESCRIPTION
**Pull request by chengseong**
**DukePro**

> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember the things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER FAST_ to use

All you need to do is,

1. download it form [here](https://nus-cs2103-ay2122s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [X] Managing tasks
- [ ] Managing deadlines (coming soon) 
- [ ] Reminders (coming soon)

------------

If you Java programmer, you can use it to practice Java too. Here's the ``main`` method :

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

### CODE REVIEW

File: src/main/java/AddCommand.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import duke.Ui;
+import duke.Storage;
+import duke.task.Task;
+
+/**
+ * Represents a Command that adds a Task to the TaskList.
+ * An AddCommand consists of a Task.
+ *
+ */
+
```
Comment (deetoomoo) [inline src/main/java/AddCommand.java]: The coding standard used for this module suggests leaving no blank line between the documentation block and the method/class. Perhaps it might save some space too?

File: src/main/java/FindCommand.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import duke.Storage;
+import duke.task.Task;
+
+/**
+ * Represents a Command that finds a Task to the TaskList.
+ * An FindCommand consists of a keyword.
+ *
+ */
+
+public class FindCommand extends Command{
```
Comment (deetoomoo) [inline src/main/java/FindCommand.java]: There appears to be a missing whitespace here after Command :-)

File: src/main/java/Task.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    /**
+     * Changes the done variable stored in Task to 1, representing a completed task.
+     */
+
+    public void complete(){
+        this.done = 1;
+    }
+
```
Comment (deetoomoo) [inline src/main/java/Task.java]: Maybe you should include javadocs for some of these public functions

**Review by deetoomoo — COMMENTED** [review]
Overall good adherence to the java coding standard! Just some minor issues.. good job!

File: data/tasks.txt [hunk 4]
```diff
@@ -0,0 +1,2 @@
+D | 1 | return book  | Dec 02 2019, 06:00 PM
```
Comment (Haozeee) [inline data/tasks.txt]: Perhaps this file should be included in the gitignore file since the tasks in it are only relevant to 1 particular user.

File: src/main/java/Parser.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+            return new DeleteCommand(toDelete);
+        case "todo": {
+            String task = fullCommand.replaceFirst("todo ", "");
+            if (task.equals("todo")) {
+                throw new IllegalTaskException("☹ OOPS!!! The description of a todo cannot be empty.");
+            } else {
+                return new AddCommand(new ToDo(task));
+            }
+        }
+        case "deadline": {
```
Comment (Haozeee) [inline src/main/java/Parser.java]: Personally, I think the case branch below is too long which makes the entire method looks unsightly and hard to read. Perhaps consider putting it into another method such as:
```java
private Command parseDeadline(String fullCommand)
```
This will make the code neater. Same thing for the Event case below.

File: src/main/java/Storage.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+     * Creates a Storage Object containing the file path of where the user wishes to store the TaskList.
+     * @param filePath file path of Storage.
+     */
+    public Storage(String filePath) {
+        this.filePath = filePath;
+    }
+
+    /**
+     * Reads and loads the storage into an ArrayList of Tasks.
+     * @return ArrayList of tasks stored in the storage.
```
Comment (Haozeee) [inline src/main/java/Storage.java]: There should be an empty line between the method and description and parameters according to the coding standard.

File: src/main/java/Storage.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+     * @throws FileNotFoundException if file path does not exist.
+     */
+
+    public ArrayList<Task> load() throws FileNotFoundException {
```
Comment (Haozeee) [inline src/main/java/Storage.java]: I think there is too much nesting in this method which makes the code hard to follow. Consider putting some of the if-else parts into seperate methods in order to simplify it.

File: src/main/java/Task.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
+package duke.task;
+
+/**
+ * Task is the abstract base class of all the Tasks.
+ */
+
+public abstract class Task {
+    private int done;
```
Comment (Haozeee) [inline src/main/java/Task.java]: Consider using a boolean instead to indicate whether a task is done or not.

File: src/main/java/ExitCommand.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+     * Executes the Exit Command.
+     *
+     * @param tasks TaskList to write into Storage.
+     * @param ui UI to reflect the exit.
+     * @param storage Storage to write the TaskList into.
+     * @throws IOException On output Error.
+     * @see IOException
+     */
+
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws IOException {
```
Comment (Haozeee) [inline src/main/java/ExitCommand.java]: There should be an override annotation here to indicate to the reader that this method is being overridden from a parent class. Same thing for the other classes that extend the Command abstract class.

File: src/main/java/TaskList.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    public TaskList(ArrayList<Task> tasks) {
+        this.tasks = tasks;
+    }
+
+    /**
+     * Deletes a task from the TaskList stored in the index given.
+     * @param index index of the task to be deleted.
+     */
+    public void delete(int index) {
+        tasks.remove(index);
```
Comment (Haozeee) [inline src/main/java/TaskList.java]: I believe this will throw an exception that is not caught if the user specifies an index that is beyond the length of the ArrayList. May want to try it out by writing a test case for it.

**Review by Haozeee — COMMENTED** [review]
Overall, I think the code is mostly easy to understand and follow. However, there are some things that can be improved. For instance, some methods can be made more concise to further improve code readability.  As for the coding standard, there should not be a space between a Javadoc comment and the method itself.
